core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2240,6 +2240,29 @@ private[zio] trait ZIOFunctions extends Serializable {
       }
     }
 
+  final def fromFutureInterrupt[A](make: ExecutionContext => scala.concurrent.Future[A]): Task[A] =
+    Task.descriptorWith { d =>
+      val ec          = d.executor.asEC
+      val interrupted = new java.util.concurrent.atomic.AtomicBoolean(false)
+      val interruptibleEC = new scala.concurrent.ExecutionContext {
+        def execute(runnable: Runnable): Unit =
+          if (!interrupted.get) ec.execute(runnable) else ()
+        def reportFailure(cause: Throwable): Unit =
+          ec.reportFailure(cause)
+      }
+      effect(make(interruptibleEC)).flatMap { f =>
+        f.value
+          .fold(
+            Task.effectAsync { (cb: Task[A] => Unit) =>
+              f.onComplete {
+                case Success(a) => cb(Task.succeed(a))
+                case Failure(t) => cb(Task.fail(t))
+              }(interruptibleEC)
+            }
+          )(Task.fromTry(_))
+      }.onInterrupt(Task.effectTotal(interrupted.set(true)))
```

Comment:
This looks great. The only thing I would suggest is to "block" interruption on the interruption of the `Future`. This will allow very precise knowledge of when the future is terminated to propagate upward (of course users can always do `.interrupt.fork` if they don't want that).

To do this, you can create an outer async effect, and complete it from inside the `interruptibleEC` (after the Future is interrupted, i.e. on failed submit), but also complete it if the Future is completed.

Does that make sense?

Reply:
I think it raises the question of when we consider the future as being interrupted.

I had been conceptualizing it as the future being interrupted when `interrupted` is set to `true`, because then we know that no further steps can occur in the chain of continuations.

It sounds like you are proposing a concept where the future counts as being interrupted when the first runnable submitted for execution is rejected.

I'm not sure that actually gives us a stronger guarantee and I think it raises some potential issues.

I think the stronger guarantee you would want is that the future is not doing anything anymore. But the original future could have spawned multiple futures so even if one runnable submitted for execution is rejected there could be other futures that are still "in flight".

The other issue is that it makes the timing of interruption dependent on the timing of the future. If we have the equivalent of `Future.never` interruption would not terminate because the future would never complete or submit to the execution context. Or if the future only continues at a long interval interruption has to wait for the next interval.

Reply:
What you say is true. Waiting for the next submit would guarantee that _some_ `Future` has been interrupted, but it's possible the implicit `ec` would have been passed and used in many `Future`'s, so we can't say it's the only continuation of the originally created `Future`. 

My only concern about the current implementation is that:

 - Interrupting a ZIO effect may never terminate: that's a risk of calling `interrupt` without also forking that interrupt
 - Interrupting a ZIO effect may not terminate all children; if it's children were created with `daemon` mode on
 - Interrupting a ZIO effect doesn't return until we know for sure it's done executing

So it seems to me there are similar risks with interrupting a `Future`. 

Also if you look at `effectBlocking`, when we interrupt some synchronous code, we actually do not return from `interrupt` until it's actually interrupted, despite the potential thread leak resulting from this behavior.

So it seems more consistent to _try_ to emulate that semantic for `Future`, together with its inherent risks. Although your point above is quite valid: at best we can only know we interrupted _some_ continuation of the `Future`, and if there are many continuations, we won't know and can't wait around for them.

Thoughts?

Reply:
I think those are all good points. This seems like one of those situations where we are interacting with code that is not principled in the same way we are so we aren't going to be able to do something that is completely principled and just have to pick the best thing. And I think you have done a good job of laying out as good a set of principles as we can have in that situation.

My only worry is that the implications of that can be counterintuitive. For example, interrupting a fiber forked from `ZIO.never` is completely safe and will return almost immediately. On the other hand, using `ZIO.fromFutureInterrupt` with `Future.never` and attempting to interrupt that will hang. That makes sense if you say we need some at least "best efforts" evidence that the Future is done executing but that can be a subtle point for two things that look similar on one level.

Totally happy to do as you suggest though.

Reply:
I changed it as you suggested. I struggled a little bit with the implementation here. I tried to use an outer asynchronous effect as you suggested but I found that I needed one effect to signal the interruption, which needed to be the "outer" effect because I needed it in scope when creating the `interruptibleEC`, but then that ended up being the result type of the composed effect.

Reply:
> My only worry is that the implications of that can be counterintuitive. 

I agree with you here. It's counterintuitive. 

The flipside is the counterintuition from the other solution: you interrupt a future but it's still doing things in the background (which means, for one: it's not safe to attempt to do the same thing in some solutions without leaking resources, e.g. consider an http request being thrown away, it is not "resource safe" to begin another one until the old one killed).

Like you say, it's a war of competing tradeoffs and I don't think any solution will be optimal, because future. On the net I believe — but do not know for sure — that an infinite future will be comparatively rare, while interruption is so automatic and composable that if the user has composed a ZIO effect from many ZIO effects and some futures, the fact that interrupting the whole chain has definitely left things running in the background due to the futures may be detectable in more situations.

Anyway, if you're happy with this, I am good to merge, looks great!

Reply:
I'm good with it.